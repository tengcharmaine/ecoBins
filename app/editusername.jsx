import { Text, Button, TextInput } from "react-native-paper";
import { StyleSheet, View, TouchableOpacity, Image, ActivityIndicator } from "react-native";
import { Link } from "expo-router";
import { useNavigation } from '@react-navigation/native';
import { createNativeStackNavigator } from '@react-navigation/native-stack';
import { useState } from "react";
import { supabase } from "../lib/supabase";
import { Ionicons } from '@expo/vector-icons';
import { useFonts } from 'expo-font';



export default function editusername() {
  const [loaded] = useFonts({
    Poppins: require('../assets/fonts/Poppins-Regular.ttf'),
    PoppinsBold: require('../assets/fonts/Poppins-Bold.ttf'),
    PoppinsSemiBold: require('../assets/fonts/Poppins-SemiBold.ttf'),
    PoppinsBlack: require('../assets/fonts/Poppins-Black.ttf'),

    // Add other Poppins variants here (e.g., Poppins-Bold, Poppins-Italic, etc.) if needed.
  });

  const styles = StyleSheet.create({
    container: {
      flex: 1,
      justifyContent: "center",
      alignItems: "center",
      backgroundColor: "white",
    },
    input: {
      borderColor: "white",
      borderWidth: 1,
      borderBottomColor: "grey",
      backgroundColor: "white",
      width: '75%',
    },
    button: {
      borderColor: "black",
      alignItems: 'center',
      justifyContent: 'center',
      backgroundColor: "#c7dede",
      width: '85%',
      height: 60,
      marginTop: 20,
      marginBottom: 10,
      borderRadius: 20,
    },
    text1: {
      color: "black",
      fontWeight: 'bold',
      fontSize: 19,
      textAlign: 'center',
      fontFamily: 'PoppinsSemiBold',
    },

    title: {
      color: "black",
      marginTop: 20,
      justifyContent: 'center',
      fontWeight: 'bold',
      fontSize: 34,
      textAlign: 'left',
      marginRight: 40,
      marginBottom: 15,
      fontFamily: 'PoppinsBold',
    },
    innerContainer: {
      flexDirection: 'row',
      alignItems: 'flex-end',
      marginBottom: 20,
    },
    emailIcon: {
      width: 30,
      height: 30,
      resizeMode: 'contain',
      marginRight: 20,
    },
    errorText: {
      color: "red",
      //marginTop: 10,
      fontFamily: 'Poppins',
    },
    backButton: {
      position: 'absolute',
      top: 60,
      left: 20,
      zIndex: 1,
    },
  });

  const [name, setname] = useState(null);
  const [error, setError] = useState(null);
  const navigation = useNavigation();

  const saveUsername = async () => {
    if (!name) {
      setError('Username cannot be empty');
      return;
    }

    try {
      const { data: { user } } = await supabase.auth.getUser();

      // Check if the new username is already taken
      const { data: existingUser } = await supabase
        .from('users')
        .select('id')
        .eq('email', name)
        .single();

      if (existingUser) {
        setError('Username is already taken. Please choose a different username.');
        return;
      }

      // Update the user's username in the Supabase table
      const { data, error } = await supabase
        .from('users')
        .update({ email: name })
        .eq('id', user.id)
        .single();

      if (error) {
        setError('Error updating username: ' + error.message);
        return;
      }

      console.log('Username updated successfully!');
      navigation.goBack();
    } catch (error) {
      setError('Error updating username: ' + error.message);
    }
  };

  const goBack = () => {
    navigation.navigate('settings');
  }

  if (!loaded) {
    return (
      <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
        <ActivityIndicator size="large" color="black" />
      </View>
    );
  }

  return (
    <View style={styles.container}>
      <TouchableOpacity onPress={goBack} style={styles.backButton}>
          <Ionicons name="arrow-back" size={24} color="black" />
      </TouchableOpacity>
      <Text style={styles.title}>Change username</Text>
      <View style= {styles.innerContainer}>
          <Image
            source={require('./../images/user-pen.png')}
            style={styles.emailIcon}
          />
          <TextInput
            autoCapitalize="none"
            placeholder="Username"
            placeholderTextColor={"#dfd8dc"}
            style={styles.input}
            textContentType="username"
            value={name}
            onChangeText={setname}
          />
      </View>
      {error && <Text style={styles.errorText}>{error}</Text>}
      <TouchableOpacity style={styles.button} onPress={saveUsername}>
        <Text style={styles.text1}>Confirm</Text>
      </TouchableOpacity>
    </View>
  );
}
