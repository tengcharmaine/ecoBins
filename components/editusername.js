import { Text, Button, TextInput } from "react-native-paper";
import { StyleSheet, View, TouchableOpacity } from "react-native";
import { useNavigation } from '@react-navigation/native';
import { useState } from "react";
import { supabase } from "../lib/supabase";
// import { Ionicons } from '@expo/vector-icons';


export default function editusername() {
  const styles = StyleSheet.create({
    container: {
      flex: 1,
      justifyContent: "center",
      alignItems: "center",
    },
    input: {
      borderColor: "black",
      borderWidth: 1,
      backgroundColor: "white",
      width: "75%",
      borderRadius: 5,
    },
    button: {
      borderColor: "black",
      alignItems: "center",
      backgroundColor: "#c7dede",
      width: "25%",
      marginTop: 20,
      borderRadius: 10,
      color: "black", 
    },
    text1: {
      color: "black",
      marginTop: 20,
      textAlign: "left",
      marginRight: 230,
      marginBottom: 5,
    },

    title: {
      color: "black",
      fontSize: 20,
      marginBottom: 20,
      fontWeight: "bold",
    },
    errorText: {
      color: "red",
      marginTop: 10,
    },
    backButton: {
      position: 'absolute',
      top: 60,
      left: 20,
      zIndex: 1,
    },
  });

  const [name, setname] = useState(null);
  const [error, setError] = useState(null);
  const navigation = useNavigation();

  const saveUsername = async () => {
    if (!name) {
      setError('Username cannot be empty');
      return;
    }

    try {
      const { data: { user } } = await supabase.auth.getUser();

      // Check if the new username is already taken
      const { data: existingUser } = await supabase
        .from('users')
        .select('id')
        .eq('email', name)
        .single();

      if (existingUser) {
        setError('Username is already taken. Please choose a different username.');
        return;
      }

      // Update the user's username in the Supabase table
      const { data, error } = await supabase
        .from('users')
        .update({ email: name })
        .eq('id', user.id)
        .single();

      if (error) {
        setError('Error updating username: ' + error.message);
        return;
      }

      console.log('Username updated successfully!');
      navigation.goBack();
    } catch (error) {
      setError('Error updating username: ' + error.message);
    }
  };

  const goBack = () => {
    navigation.navigate('settings');
  }

  return (
    <View style={styles.container}>
      <TouchableOpacity onPress={goBack} style={styles.backButton}>
          {/* <Ionicons name="arrow-back" size={24} color="black" /> */}
      </TouchableOpacity>
      <Text style={styles.title}>Change username</Text>
      <TextInput
        autoCapitalize="none"
        placeholder="Username"
        placeholderTextColor={"#dfd8dc"}
        style={styles.input}
        textContentType="username"
        value={name}
        onChangeText={setname}
      />
      {error && <Text style={styles.errorText}>{error}</Text>}
      <Button style={styles.button} onPress={saveUsername}>
        <Text style={styles.text1}>Confirm</Text>
      </Button>
    </View>
  );
}