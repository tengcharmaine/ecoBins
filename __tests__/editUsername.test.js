import React from 'react';
import renderer from 'react-test-renderer';
import editusername from '../components/editusername';


describe('Register', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });
  test('renders correctly', () => {
    const tree = renderer.create(<editusername />).toJSON();
    
    // Check if the required elements are rendered
    expect(tree).toMatchSnapshot();
  });

  test('renders correctly with no errors and initial value', () => {
    // Set an initial value for the name state
    const initialName = 'initialUsername';
  
    // Render the component with the initialName as a prop
    const tree = renderer.create(<editusername name={initialName} />).toJSON();
  
    // Check if the required elements are rendered
    expect(tree).toMatchSnapshot();
  });

  test('renders correctly with an error message', () => {
    // Set an error message for the error state
    const errorMessage = 'Username is already taken';
  
    // Render the component with the errorMessage as a prop
    const tree = renderer.create(<editusername error={errorMessage} />).toJSON();
  
    // Check if the required elements are rendered, including the error message
    expect(tree).toMatchSnapshot();
  });

});