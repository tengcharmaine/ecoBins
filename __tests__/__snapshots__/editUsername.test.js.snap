// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register renders correctly 1`] = `<editusername />`;

exports[`Register renders correctly with an error message 1`] = `
<editusername
  error="Username is already taken"
/>
`;

exports[`Register renders correctly with no errors and initial value 1`] = `
<editusername
  name="initialUsername"
/>
`;
